### PR TITLE
Pass snippets as argument to engine

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
Makes it so that Cursorless everywhere editors doesn't have to provide a snippet implementation

## Checklist

- [/] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [/] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [/] I have not broken the cheatsheet

### CODE REVIEW

**Review by pokey — COMMENTED** [review]
I haven't read in detail, but my first two reactions are as follows:

Moving snippets from engine to vscode feels like a lateral at best. We went through the effort to abstract vscode out of the snippets component, so I would prefer if we didn't reverse that. Ie this makes it harder for us to use snippets in neovim or other node contexts

Second issue is that it doesn't really make the core much leaner if we still have a bunch of references to snippets in the engine. They're only used in two actions right? I'd be tempted to think the right abstraction is to allow passing in custom actions. Then snippets are never even mentioned

But I didn't read super closely so maybe missing something

**Comment by AndreasArvidsson** [comment]
I've had another look and started a more more abstracted implementation and bailed. Basically most of what is in `Snippets.ts` that I here reamed to `VscodeSnippets.ts` is vscode specific. The fact that we are polling the file system. The handling of directory error message. The fact that you can a register third party snippets to some extent. If I remove everything that isn't vscode and make an abstraction it basically becomes more code for no good reason. I stand by my original assessment that this should probably live in vscode until we make a refactor where we used the new snippet format and don't poll.

**Comment by AndreasArvidsson** [comment]
Implementations like this also makes it a bit hard to abstract snippets. We are explicitly creating a file in the file system and then opening in the editor. To open the file we need a filepath so we are still stuck in the file system with snippets.
https://github.com/cursorless-dev/cursorless/blob/2e69df8b9b8852a1df6cba5a6b5b7a73d82d6c63/packages/cursorless-engine/src/actions/GenerateSnippet/openNewSnippetFile.ts#L20

**Comment by pokey** [comment]
> Implementations like this also makes it a bit hard to abstract snippets. We are explicitly creating a file in the file system and then opening in the editor. To open the file we need a filepath so we are still stuck in the file system with snippets. Here my inclination is just to drill the file system instance through and add touch to that. We can later on revamp the entire snippet system and make its smarter, but for now I can a just need to be unblocked. Sounds reasonable?
> 
> https://github.com/cursorless-dev/cursorless/blob/2e69df8b9b8852a1df6cba5a6b5b7a73d82d6c63/packages/cursorless-engine/src/actions/GenerateSnippet/openNewSnippetFile.ts#L20

Wouldn't my second suggestion above solve this problem? Ie generate snippet, insert snippet and wrap with snippet are just custom actions passed into the engine. If this isn't making sense let's discuss tomorrow alongside vscode vs engine thing

**Comment by AndreasArvidsson** [comment]
That sounds like a reasonable solution, but to be fair I don't really have the time for that right now. This was only moving a single file that will unblock me. And we can then later make a more flexible approach. Happy to discuss it tomorrow.

**Comment by pokey** [comment]
Let's discuss tomorrow

**Comment by pokey** [comment]
update from meet-up: pokey caved

**Review by pokey — COMMENTED** [review]
The one thing I think we discussed changing here is creating a `DisabledSnippets` class so that we can confine null checks to `createEngine`. Otherwise looks good; see also https://github.com/cursorless-dev/cursorless/issues/2353#issuecomment-2222821726

File: packages/cursorless-engine/src/cursorlessEngine.ts [hunk 1]
```diff
@@ -30,24 +31,22 @@ import { ScopeSupportChecker } from "./scopeProviders/ScopeSupportChecker";
...
-import { KeyboardTargetUpdater } from "./KeyboardTargetUpdater";
+import { DisabledSnippets } from "./disabledComponents/DisabledSnippets";
 
 export async function createCursorlessEngine(
   treeSitter: TreeSitter,
   ide: IDE,
   hats: Hats,
   commandServerApi: CommandServerApi | null,
   fileSystem: FileSystem,
+  snippets: Snippets | undefined = new DisabledSnippets(),
```
Comment (pokey) [inline packages/cursorless-engine/src/cursorlessEngine.ts:42]:
```suggestion
  snippets: Snippets = new DisabledSnippets(),
```

**Review by pokey — COMMENTED** [review]
Looks good with minor nit

**Review by pokey — APPROVED** [review]

**Pull request merged**